### PR TITLE
[TOPIC-GPIO] drivers: ht16k33: update to use new GPIO API

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Update the HT16K33 LED driver, key-scan driver, and sample application to use the new GPIO API.

### CODE REVIEW

File: samples/drivers/ht16k33/nrf52840_pca10056.overlay [hunk 1]
```diff
@@ -12,7 +12,7 @@
 		reg = <0x70>;
 		label = "HT16K33";
                 /* Uncomment to use IRQ instead of polling: */
-		/* irq-gpios = <&gpio1 8 (GPIO_PUD_PULL_UP | GPIO_INT_ACTIVE_LOW)>; */
+		/* irq-gpios = <&gpio1 8 GPIO_PUD_PULL_UP>; */
```
Comment (pabigot) [inline samples/drivers/ht16k33/nrf52840_pca10056.overlay]: The (undocumented) consensus seems to be that active level should be made explicit, so this should add `GPIO_ACTIVE_HIGH`.  Though that's not consistent with normal use of pull-up, so maybe it should be `GPIO_ACTIVE_LOW`.  Except in the driver the interrupt is configured for edge-to-inactive.

File: drivers/gpio/gpio_ht16k33.c [hunk 2]
```diff
@@ -36,14 +36,14 @@ struct gpio_ht16k33_data {
...
-			       u32_t pin, int flags)
+			    u32_t pin, int flags)
 {
 	ARG_UNUSED(dev);
 	ARG_UNUSED(access_op);
 	ARG_UNUSED(pin);
 
 	/* Keyscan is input-only */
-	if ((flags & GPIO_DIR_MASK) != GPIO_DIR_IN) {
+	if ((flags & GPIO_INPUT) != GPIO_INPUT) {
```
Comment (pabigot) [inline drivers/gpio/gpio_ht16k33.c]: This looks a little weird.  I assume this device only supports input, no output; and that it doesn't support disconnecting the input.  If so, I think this should be:
```
if (((flags & (GPIO_INPUT | GPIO_OUTPUT)) == GPIO_DISCONNECTED)
    || ((flags & GPIO_OUTPUT) != 0)) {
    return -ENOTSUP;
}
```

**Review by pabigot — CHANGES_REQUESTED** [review]
Fix needed for validation of the supported I/O modes for the GPIO.

Also it's very confusing whether the interrupt is being configured for a logic level or a physical level.  The devicetree uses a pull-up, and the interrupt seems to detect transition to low, but the devicetree defaults to active high.  I would normally expect that pull configuration to indicate an active-low trigger on edge-to-active.

**Comment by mnkp** [comment]
HT16K33 is not a General Purpose Input Output driver. I think it would be cleaner to move all the interrupt handling code directly into `drivers/led/ht16k33.c` and remove `gpio_ht16k33.c` driver.

**Comment by pabigot** [comment]
> HT16K33 is not a General Purpose Input Output driver. I think it would be cleaner to move all the interrupt handling code directly into `drivers/led/ht16k33.c` and remove `gpio_ht16k33.c` driver.

I agree with this in principle, especially as it appears that there's additional programming necessary to configure and access the keypresses.

Since the LED driver doesn't support a keypress callback API taking that approach would require either a driver-specific API for the ht16k33 or an extension to the LED API to optionally support key presses (if this is not a unique feature of this controller).

For now, since it's already in-tree, I don't think its unreasonable to just make it work the way it has been.

**Comment by carlescufi** [comment]
@henrikbrixandersen 
The topic branch has been rebased. Could you please rebase this PR against it?

File: drivers/gpio/gpio_ht16k33.c (hunk 2 above)
Comment (henrikbrixandersen) [reply drivers/gpio/gpio_ht16k33.c]: @pabigot Thanks, I have changed it accordingly.

File: samples/drivers/ht16k33/nrf52840_pca10056.overlay (hunk 1 above)
Comment (henrikbrixandersen) [reply samples/drivers/ht16k33/nrf52840_pca10056.overlay]: @pabigot My intention was to to use an IRQ which triggers on falling edges. I have updated it to explicitly specify the active level.

**Comment by henrikbrixandersen** [comment]
> Since the LED driver doesn't support a keypress callback API taking that approach would require either a driver-specific API for the ht16k33 or an extension to the LED API to optionally support key presses (if this is not a unique feature of this controller).

I have been meaning to introduce a simple key event input subsystem in Zephyr for a while. For now, the closest we have seems to be the GPIO interrupt + callback functionality, which is why the HT16K33 driver is implementing that API.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen
> The topic branch has been rebased. Could you please rebase this PR against it?

@carlescufi Rebased and fixed (I hope) the issues brought up in the review so far.

**Review by pabigot — APPROVED** [review]
I think this is OK, though of course I can't test it.

**Pull request merged**